### PR TITLE
chore(asset-service): add colors from icons

### LINKED ISSUE #770: Asset service - color detection + overrides
### Overview

Motivation - rainbow dashboard portfolio charts.

We currently have a color field on icons - but they're mostly set as white.

We want to be able to automatically detect this from icons, and populate a file, and additionally have an overrides file, where we can manually set accurate colors for icons.

### References and additional details

* https://github.com/shapeshift/lib/blob/main/packages/asset-service/src/generateAssetData/generateAssetData.ts#L34-L35
* We want a separate script, something like `yarn generate-colors` in the asset service
* The script should, take the generated asset data, fetch the icons for each asset (respecting rate limiting), use some library/algo to determine the mode color (most common), and output the hex value (should this be RGBA?) to a json file called `colors.json`, which is keyed by asset id, and the value is the color.
* Additionally, we should have a `colors-overrides.json` with the same structure as the `colors.json` file where we can manually set more accurate colors for popular/requested assets. This should not be modified by scripts.
* The script should run independently of generating asset data, as it will be slow and colors don't change often.

### Acceptance Criteria

* Some script command we can run independently to automatically generate colors
* The override files describe above
* Generated asset data with fancy pants colors

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

1 day

### PR DESCRIPTION
**Pull request by theoboldfrazier**
This PR creates a color map that grabbed the dominant (most prevalent) color from the asset's icon.

It then uses the color map to set the colors in the assets generation script. If there is a new asset not found in the map, it will create try to create a color from the icon and set that color.

I created a map here, because the initial run of checking the color of each remote icon took about 20 min for the ~6000 assets. With the map, there is negligible addition to the time of running the asset generation script.

Closes #770

### CODE REVIEW

File: packages/asset-service/src/generateAssetData/setColors.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  return hex.length === 1 ? `0${hex}` : hex
+}
+
+export const setColors = async (assets: Asset[]): Promise<Asset[]> => {
+  for (let i = 0; i < assets.length - 1; i++) {
+    try {
+      if (assets[i].color === '#FFFFFF' && assets[i].icon) {
+        // colorThief.getColor returns the most dominant color in the icon.
+        const color = await colorThief.getColor(assets[i].icon)
+        const hexColor = `#${toHex(color[0])}${toHex(color[1])}${toHex(color[2])}`
```
Comment (gomesalexandre) [inline packages/asset-service/src/generateAssetData/setColors.ts:15]: Non-blocking, I'd name these `red` `green` and `blue` (or `r`, `g` and `b`) but given we're probably not going to change that script's logic, I'm fine with this

File: packages/asset-service/src/generateAssetData/generateColorMap.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import 'dotenv/config'
+
+import { AssetId } from '@shapeshiftoss/caip'
+import { Asset } from '@shapeshiftoss/types'
+import fs from 'fs'
```
Comment (gomesalexandre) [inline packages/asset-service/src/generateAssetData/generateColorMap.ts:5]: Might or might not tree-shake better
```suggestion
import fs from `fs/promises'
```

**Review by gomesalexandre — APPROVED** [review]
LGTM - noticed a few rugs in `generateColorMap` on assets such as AAVE LEND - old assets for which trustwallet/assets doesn't have logos, and this is gracefully handled. e.g:

https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0x80fB784B7eD66730e8b1DBd9820aFD29931aab03/logo.png
https://github.com/trustwallet/assets/tree/master/blockchains/ethereum/assets/0x80fB784B7eD66730e8b1DBd9820aFD29931aab03

File: packages/asset-service/src/generateAssetData/setColors.ts (hunk 1 above)
Comment (theoboldfrazier) [reply packages/asset-service/src/generateAssetData/setColors.ts]: Good call. Destructuring makes sense.

File: packages/asset-service/src/generateAssetData/generateColorMap.ts (hunk 2 above)
Comment (theoboldfrazier) [reply packages/asset-service/src/generateAssetData/generateColorMap.ts:5]: Since it's a local script, I just kept as fs.

**Review by 0xApotheosis — COMMENTED** [review]
Nice one, some non-blocking suggestions.

I am surprised though at the diff I got when running the script, maybe network failures resulted in reverting to the defaults?

Screen Shot 2022-06-17 at 10 08 54 am

File: packages/asset-service/src/generateAssetData/setColors.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+function toHex(num: number): string {
+  const hex = num.toString(16).toUpperCase()
+  return hex.length === 1 ? `0${hex}` : hex
+}
+
+export const setColors = async (assets: Asset[]): Promise<Asset[]> => {
+  for (let i = 0; i < assets.length - 1; i++) {
+    try {
+      if (assets[i].color === '#FFFFFF' && assets[i].icon) {
```
[the remark this answers is not in the document]
Comment (theoboldfrazier) [reply packages/asset-service/src/generateAssetData/setColors.ts:9]: The original list had all of them set as #FFFFFF, so this conditional was used to identify colors that had the default.

These changes don't specifically set any to #FFFFFF, they use that value on the existing assets to determine if they had been set.

File: packages/asset-service/src/generateAssetData/generateColorMap.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  const ethAssets = await addTokensToEth()
+  const osmosisAssets = await getOsmosisAssets()
+
+  // all assets, included assets to be blacklisted
+  const unfilteredAssetData: Asset[] = [bitcoin, tBitcoin, ...ethAssets, atom, ...osmosisAssets]
+  // remove blacklisted assets
+  const filteredAssetData = filter(
+    unfilteredAssetData,
+    ({ assetId }) => !blacklist.includes(assetId)
+  )
```
[the remark this answers is not in the document]
Comment (theoboldfrazier) [reply packages/asset-service/src/generateAssetData/generateColorMap.ts]: Done.

File: packages/asset-service/src/generateAssetData/setColors.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { Asset } from '@shapeshiftoss/types'
+import colorThief from 'colorthief'
+
+function toHex(num: number): string {
+  const hex = num.toString(16).toUpperCase()
+  return hex.length === 1 ? `0${hex}` : hex
+}
+
+export const setColors = async (assets: Asset[]): Promise<Asset[]> => {
+  for (let i = 0; i < assets.length - 1; i++) {
```
[the remark this answers is not in the document]
Comment (theoboldfrazier) [reply packages/asset-service/src/generateAssetData/setColors.ts]: Done

File: packages/asset-service/src/generateAssetData/setColors.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply packages/asset-service/src/generateAssetData/setColors.ts]: Sure, we might wanna make that type change in the future - fine to leave out of this PR.

**Comment by theoboldfrazier** [comment]
> Nice one, some non-blocking suggestions.
> 
> I am surprised though at the diff I got when running the script, maybe network failures resulted in reverting to the defaults?
> 
> Screen Shot 2022-06-17 at 10 08 54 am

Fixed in this commit: https://github.com/shapeshift/lib/pull/826/commits/4ee0091361196fb782a582dccae02cef63f3d2d5

**Review by cjthompson — APPROVED** [review]

**Review by 0xApotheosis — APPROVED** [review]
Tested the changes, no diff with the generated output 👌 

🍭

**Pull request merged**